topdown/http.go

```diff
@@ -275,34 +293,82 @@ func executeHTTPRequest(bctx BuiltinContext, obj ast.Object) (ast.Value, error)
 		tlsConfig.InsecureSkipVerify = tlsInsecureSkipVerify
 	}
 
+	if len(tlsClientCert) > 0 && len(tlsClientKey) > 0 {
+		tlsClientCert = bytes.Replace(tlsClientCert, []byte("\\n"), []byte("\n"), -1)
+		tlsClientKey = bytes.Replace(tlsClientKey, []byte("\\n"), []byte("\n"), -1)
```

Comment:
This seems kind of odd, why do we need to do this? Is the expectation that someone would provide escaped newlines?

For reference, I don't think we have that behavior for certificates provided to the `crypto.x509.parse_certificates()` or `io.jwt.*` builtins when they pass in x.509 certificates. If it is something we should be doing in all of these spots we should file some issues to track that and get them all consistent.

Reply:
> This seems kind of odd, why do we need to do this? Is the expectation that someone would provide escaped newlines?

Yeh, it's super weird. It looks to me like the Rego interpreter is taking a raw string and turning it into a Go literal. But it's also possible that I'm missing a layer of quoting somewhere :)

I'll look at the tests for the builtins you mention and see whether I can emulate that.

Reply:
@patrick-east So `crypto.x509.parse_certificates` wants a ASN.1 input, so you have to give it a base64 string. Not clear to me what you should do if you have PEM.

The JWT tests end up using single-quoted strings with `\n` characters.

The TLS test cases here end up formatting Rego that looks like [this](https://play.openpolicyagent.org/p/VYtV7hLCKV). I verified that Rego does actually convert the newlines to a literal `\n` somewhere:

```
--- topdown/http.go
+++ topdown/http.go
@@ -12,6 +12,7 @@ import (
        "fmt"
        "io"
        "io/ioutil"
+       "log"
        "net/http"
        "net/url"
        "os"
@@ -294,8 +295,14 @@ func executeHTTPRequest(bctx BuiltinContext, obj ast.Object) (ast.Value, error)
        }

        if len(tlsClientCert) > 0 && len(tlsClientKey) > 0 {
+               log.Printf("CA Cert ---")
+               log.Printf(string(tlsClientCert))
+               log.Printf("CA Cert ---")
                tlsClientCert = bytes.Replace(tlsClientCert, []byte("\\n"), []byte("\n"), -1)
                tlsClientKey = bytes.Replace(tlsClientKey, []byte("\\n"), []byte("\n"), -1)
+               log.Printf("CA Cert ---")
+               log.Printf(string(tlsClientCert))
+               log.Printf("CA Cert ---")
                cert, err := tls.X509KeyPair(tlsClientCert, tlsClientKey)
                if err != nil {
                        return nil, err
```

This diff produces the following tests output:
```
=== RUN   TestHTTPSClient/HTTPS_Get_with_Inline_Cert/http.send
2020/04/07 14:24:19 CA Cert ---
2020/04/07 14:24:19 -----BEGIN CERTIFICATE-----\nMIICoDCCAYgCCQDDVCvl5ehtCTANBgkqhkiG9w0BAQUFADAQMQ4wDAYDVQQDDAVt\neS1jYTAeFw0xOTAxMjcxOTAzNTNaFw0yMTEwMjMxOTAzNTNaMBQxEjAQBgNVBAMM\nCW15LWNsaWVudDCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBANWGICix\nhheANy6ygyATbzuCge1VXF0HgKwIm1tRmcHosu3OrLu5/IO6ci0tLH98Kh+CghVo\nbfgedg0fkarWs4XJXvHCTrXxqGIYfKo3Lsm+hL3pAq6pb3UI276urvcnNK7rujYF\nCIRGXPBSWPbfwR1fqMu01ea3LlhNOQs6jDDubSCI35rXK9nXPMUJEAgI6Cb0qBpI\nONMusytYAJCQttHfD62GcQvhcX38QitjyzCCkogfdf4kbXbEG1TVI+CwKr0L2ZId\nJeEhJq/82gHSKCNh0N7Bki+SXySndMCQv0zwNrZT9Q96Ga1d/NBBDmsbx5IWN7e+\nuwCtPpl8GXmGspsCAwEAATANBgkqhkiG9w0BAQUFAAOCAQEAK+GD1PlATdf8dk8F\nqKRMzZ3a+N5yUJH81bzFIIaarBzkVlj9Ox17dk9mf7Uxr5K7USjLyesknLTTjXgE\nfanDtTO3Pp8vcUcMO5sdKcPLdaITpKIYT+GeYiER/2lk5uSCAfMhJiP15N8HVNG3\nK1HiHu9/mxLPhQvIkNiJhuGgzg/WDg+fX+UXvRrOJtN8OBJwmTePwdc8w1JtjA7Z\n7tZJTxICcVkKJn6c9LxKdnFDPzHEDX68QBexOpTBMp64tYMWHviY75nJf3jWfUGA\nwUkU+qStgbWMQ4iXaROAk7WaBjff6A1dq7g/xlzlkSpRb/zfOE/8QWSXFOS28zG+\nMQXc9w==\n-----END CERTIFICATE-----\n
2020/04/07 14:24:19 CA Cert ---
2020/04/07 14:24:19 CA Cert ---
2020/04/07 14:24:19 -----BEGIN CERTIFICATE-----
MIICoDCCAYgCCQDDVCvl5ehtCTANBgkqhkiG9w0BAQUFADAQMQ4wDAYDVQQDDAVt
eS1jYTAeFw0xOTAxMjcxOTAzNTNaFw0yMTEwMjMxOTAzNTNaMBQxEjAQBgNVBAMM
CW15LWNsaWVudDCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBANWGICix
hheANy6ygyATbzuCge1VXF0HgKwIm1tRmcHosu3OrLu5/IO6ci0tLH98Kh+CghVo
bfgedg0fkarWs4XJXvHCTrXxqGIYfKo3Lsm+hL3pAq6pb3UI276urvcnNK7rujYF
CIRGXPBSWPbfwR1fqMu01ea3LlhNOQs6jDDubSCI35rXK9nXPMUJEAgI6Cb0qBpI
ONMusytYAJCQttHfD62GcQvhcX38QitjyzCCkogfdf4kbXbEG1TVI+CwKr0L2ZId
JeEhJq/82gHSKCNh0N7Bki+SXySndMCQv0zwNrZT9Q96Ga1d/NBBDmsbx5IWN7e+
uwCtPpl8GXmGspsCAwEAATANBgkqhkiG9w0BAQUFAAOCAQEAK+GD1PlATdf8dk8F
qKRMzZ3a+N5yUJH81bzFIIaarBzkVlj9Ox17dk9mf7Uxr5K7USjLyesknLTTjXgE
fanDtTO3Pp8vcUcMO5sdKcPLdaITpKIYT+GeYiER/2lk5uSCAfMhJiP15N8HVNG3
K1HiHu9/mxLPhQvIkNiJhuGgzg/WDg+fX+UXvRrOJtN8OBJwmTePwdc8w1JtjA7Z
7tZJTxICcVkKJn6c9LxKdnFDPzHEDX68QBexOpTBMp64tYMWHviY75nJf3jWfUGA
wUkU+qStgbWMQ4iXaROAk7WaBjff6A1dq7g/xlzlkSpRb/zfOE/8QWSXFOS28zG+
MQXc9w==
-----END CERTIFICATE-----
2020/04/07 14:24:19 CA Cert ---
```

So, given a raw string with embedded newlines, it looks like Rego turns that into a Go string with `\n` characters by the time the builtin gets it?

Reply:
Whew ok.. had to do some sifting around with the debugger to figure out what was going on, but I think we can make this work as expected.

When we use the stringer via `obj.Get(val).String()` it ends up quoting the underlying string value https://github.com/open-policy-agent/opa/blob/b39b34760e52a8db723423c89961328ff85b8612/ast/term.go#L700-L702

What we actually want is the *real* AST string value for the term, which is just the StringTerm's `Value` https://github.com/open-policy-agent/opa/blob/b39b34760e52a8db723423c89961328ff85b8612/ast/term.go#L663-L669

Up above in the big switch statement we do this already for the `raw_body`, which even checks to make sure its _really_ a string (which in retrospect all of these should be doing too).

So, if you change to doing something like:

```go
		case "tls_ca_cert":
			s, ok := obj.Get(val).Value.(ast.String)
			if !ok {
				return nil, builtins.ErrOperand("tls_ca_cert must be a string")
			}
			tlsCaCert = []byte(s)
			tlsCaCert = bytes.Trim(tlsCaCert, "\"")
```
You will get the original bytes for the certificate and can remove the check/swap for escaped newlines.

We should probably do this for all the cert options that want a string.

Reply:
Thanks for debugging this @patrick-east!

Reply:
I added a new commit that checks and converts string parameters. I removed the weirdo `\n` replacement :)